packages/falter-berlin-hna-list/src/libhna/hna.c

```diff
@@ -0,0 +1,169 @@
+                return NULL;
+        }
+
+        if (connect(sock, (struct sockaddr *) &server_addr, sizeof(server_addr)) < 0) {
+                printf("connect(): %s\n", strerror(errno));
+                return NULL;
+        }
+
+        send(sock, request, strlen(request), 0);
+        int valread = read(sock, buffer, buffer_len - 1);
+        if (valread > 0) {
+                return buffer;
+        } else {
+                return NULL;
+        }
+}
+
+
+hna_data *serialize_hna_string(char *hna, char *gateway) {
+        // input-data: "10.230.198.192/28", "10.31.43.188"
+        hna_data* dataset = calloc(1, sizeof(hna_data));
+
+        char* ipaddr = strtok(hna, "/");
+        char* netmask = strtok(NULL, "/");
+
+        //ToDo: struct sockaddr_in* addr_struct = (struct sockaddr_in)dataset->hna.base_addr;
+        inet_pton(AF_INET, ipaddr, &dataset->hna.base_addr.sin_addr);
+        dataset->hna.netmask = netmask;
+
+        inet_pton(AF_INET, gateway, &dataset->via_gateway.sin_addr);
+
+        return dataset;
+}
+
+void read_hna_into_tree(struct avl_table *tree, char *raw_data) {
+        char *str1, *str2, *token, *subtoken;
+        char *saveptr1, *saveptr2;
+        char *line_delim = "\n";
+        char *field_delim = "\t";
+        int j;
+        for (str1 = raw_data;;str1 = NULL) {
+                token = strtok_r(str1, line_delim, &saveptr1);
+                if (token == NULL) {
+                        break;
+                }
+                if (token[0] == 'T' || token[0] == 'D') {
+                        continue;
+                }
+
+                char* args[2];
+                for (j=0,str2 = token;;str2 = NULL, j++) {
+                        subtoken = strtok_r(str2, field_delim, &saveptr2);
+                        if (subtoken == NULL)
+                                break;
+                        args[j] = subtoken;
+                }
+
+                hna_data *set = serialize_hna_string(args[0], args[1]);
+                avl_insert(tree, set);
+        }
+}
+
+void read_hosts_into_tree(struct avl_table *tree, char *raw_data) {
+        char *str1, *str2, *token, *subtoken;
+        char *saveptr1, *saveptr2;
+        char *line_delim = "\n";
+        char *field_delim = "\t";
+        int j;
+        for (str1 = raw_data;;str1 = NULL) {
+                token = strtok_r(str1, line_delim, &saveptr1);
+                if (token == NULL)
+                        break;
+                if (token[0] == '#')
+                        continue;
+
+                char* args[2];
+                for (j=0,str2 = token;;str2 = NULL, j++) {
+                        subtoken = strtok_r(str2, field_delim, &saveptr2);
+                        if (subtoken == NULL)
+                                break;
+                        args[j] = subtoken;
+                }
+
+                // search-item needs to hold GW-ip-address in binary-form only
+                hna_data search;
```

Comment:
I think we need to initialize the variabe:

hna_data search = {0};

Reply:
should be resolved in the latest commit.